src/lib/fetcher/wsFetcher.ts

```diff
@@ -0,0 +1,138 @@
+import { Logger } from 'log4js';
+import { RequestConfig, Response } from '../../dex-helper/irequest-wrapper';
+import {
+  connection as WebSocketConnection,
+  client as WebSocketClient,
+} from 'websocket';
+
+export class SkippingRequest {
+  constructor(public message = '') {}
+}
+
+export type RequestInfo<T> = {
+  requestFunc?: (
+    options: RequestConfig,
+  ) => Promise<Response<T> | SkippingRequest>;
+  requestOptions: RequestConfig;
+  caster: (data: unknown) => T;
+  authenticate?: (options: RequestConfig) => RequestConfig;
+  excludedFieldsCaching?: string[];
+};
+
+export type RequestInfoWithHandler<T> = {
+  info: RequestInfo<T>;
+  handler: (data: T) => void;
+};
+
+export class WebSocketFetcher<T> {
+  private requests: RequestInfoWithHandler<T>;
+  public lastFetchSucceeded: boolean = false;
+  private stop: boolean = true;
+  private ws: WebSocketClient = new WebSocketClient();
+  private connection: WebSocketConnection | null = null;
+
+  constructor(requestsInfo: RequestInfoWithHandler<T>, private logger: Logger) {
+    this.requests = requestsInfo;
+    this.ws.on('connect', this.connected.bind(this));
+    this.ws.on('connectFailed', this.connectFailed.bind(this));
+  }
+
+  private connected(connection: WebSocketConnection) {
+    this.connection = connection;
+    this.logger.info(`Connected to ${this.requests.info.requestOptions.url}`);
+    this.connection.on('error', this.onError.bind(this));
+    this.connection.on('close', this.onClose.bind(this));
+    this.connection.on('message', this.onMessage.bind(this));
+  }
+
+  private connectFailed(error: any) {
+    this.logger.error(`Connect Error: ${error.toString()}. Reconnecting...`);
+    // reconnect on errors / failures
+    setTimeout(() => {
+      this.startPolling();
+    }, 3000);
+  }
+
+  private onClose() {
+    this.logger.info(`Connection closed. Reconnecting...`);
+    // reconnect on errors / failures
+    setTimeout(() => {
+      this.startPolling();
+    }, 3000);
+  }
+
+  private onError(error: any) {
+    this.logger.error(`Connection Error: ${error.toString()}. Reconnecting...`);
+    // reconnect on errors / failures
+    setTimeout(() => {
+      this.startPolling();
+    }, 3000);
```

Comment:
Does prev connection close before we start new one? it could leak to too many connections

Reply:
added close before trying to reconnect